floodsub.go

```diff
@@ -324,6 +359,94 @@ func msgID(pmsg *pb.Message) string {
 	return string(pmsg.GetFrom()) + string(pmsg.GetSeqno())
 }
 
+// pushMsg pushes a message to a number of subscriptions, performing validation
+// as necessary
+func (p *PubSub) pushMsg(subs []*Subscription, src peer.ID, msg *Message) {
+	// we perform validation if _any_ of the subscriptions has a validator
+	// because the message is sent once for all topics
+	needval := false
+	for _, sub := range subs {
+		if sub.validate != nil {
+			needval = true
+			break
+		}
+	}
+
+	if needval {
+		// validation is asynchronous and globally throttled with the throttleValidate semaphore
+		select {
+		case p.validateThrottle <- struct{}{}:
+			go func() {
+				p.validate(subs, src, msg)
+				<-p.validateThrottle
+			}()
+		default:
+			log.Warningf("message validation throttled; dropping message from %s", src)
```

Comment:
We'll probably want per-peer or per-topic throttles in the future but, for now, this is probably fine.

Reply:
we already have those! That was part of the improvements in this PR (although most of the stuff got rewritten).

Reply:
I see... I'm worried that it's still possible to starve a topic by maxing out the global validator first. Ideally, we'd take the per-topic ticket first, then the global one.

Reply:
I think it would be hard to do, because a throttled subscription would release the semaphore quickly.

Reply:
I improved the comments around the throttle to more clearly describe the intention and existence of per-topic throttles.